src/HeapJobOrderer.ts

```diff
@@ -0,0 +1,156 @@
+	 * @param jobs - The jobs to manage. They must have correctly populated prerequisites fields.
+	 */
+	constructor(rootJobs: Job[]) {
+		// Initialize the sources heap with its sorting function.
+		this.sourcesHeap = new Heap((job1, job2) => {
+			const job1Dependents = this.jobToDependents.get(job1)
+			const job2Dependents = this.jobToDependents.get(job2)
+
+			assert(
+				job1Dependents && job2Dependents,
+				'Provided jobs are missing from map.',
+			)
+
+			return job2Dependents.size - job1Dependents.size
+		})
+
+		// Add all of the root jobs to dependents map.
+		for (const rootJob of rootJobs) {
+			this.jobToDependents.set(rootJob, new Set())
+		}
+
+		const queue: Job[] = rootJobs
+		const seenSources: Set<Job> = new Set() // We use this because checking whether a heap has an element is O(n).
+
+		// BFS traverse dependency graph, starting with root nodes (the ultimate jobs).
+		while (queue.length) {
+			const job = queue.pop()
+			assert(job, 'Queue with nonzero length should have returned object.')
+
+			// Don't process the same node twice.
+			if (!seenSources.has(job) && !this.nonSources.has(job)) {
+				// Update job's prerequisites' 'dependents' fields.
+				for (const prerequisite of job.getPrerequisitesIterable()) {
+					let dependents = this.jobToDependents.get(prerequisite)
+
+					if (!dependents) {
+						dependents = new Set()
+						this.jobToDependents.set(prerequisite, dependents)
+					}
+
+					dependents.add(job)
+					queue.push(prerequisite)
+				}
+
+				if (job.isSource()) {
+					this.sourcesHeap.push(job)
+					seenSources.add(job)
+				} else {
+					this.nonSources.add(job)
+				}
+
+				// Initialization.
+				this.jobToCompletedPrereqs.set(job, new Set())
+			}
+		}
+	}
+
+	/**
+	 * Chooses the next job to run and returns it. Marks that job as in-progress.
+	 * If there is no job that can be run, returns null.
+	 *
+	 * @returns The next job to run.
+	 */
+	public popNextJob(): Job | null {
+		const nextJob = this.sourcesHeap.pop()
+
+		if (nextJob) {
+			this.inProgress.add(nextJob)
+		}
+
+		return nextJob || null
+	}
+
+	/**
+	 * Marks a job as completed.
+	 * JobOrderer will update the dependency graph and then forget about this job.
+	 *
+	 * @param completedJob - The job that is completed.
+	 */
+	public reportCompletedJob(completedJob: Job): void {
+		const dependents = this.jobToDependents.get(completedJob)
+
+		if (!dependents) {
+			throw new Error(
+				`We don't know about this job marked completed: ${completedJob.toString()}.`,
+			)
+		}
+
+		for (const dependent of dependents) {
+			const dependentCompletedJobs = this.jobToCompletedPrereqs.get(dependent)
+
+			assert(
+				dependentCompletedJobs,
+				`Job ${completedJob.toString()} has dependent that we didn't process, which should be impossible.`,
+			)
+			dependentCompletedJobs.add(completedJob)
+
+			if (dependent.getNumPrerequisites() === dependentCompletedJobs.size) {
```

Comment:
Add an assertion that completedJob is a member of dependent's prerequisites. Otherwise it could take a while to convince yourself that this line is equivalent to "dependent's prereqs and dependentCompletedJobs are the same set"

Reply:
The only way that job A can become a dependent of job B is if job B is in job A's prerequisites. I can add the `assert`, but isn't that kind of trivially true?

Reply:
I think my original thinking was that using a set would prevent any errors if someone called `reportCompletedJob()` twice with the same Job. But I can just move the `inProgress()` existence-check up to prevent this.

So yeah I'll just store the number. Great catch.